### PR TITLE
[Pallas] Integrate FlashAttention with SPMD

### PR DESCRIPTION
**Pull request by alanwaketan**
Summary:
This pull request integrating FlashAttention with SPMD. The way it works is to create a manual sharding region for the kernel which means we wraps all the inputs with enable_manual_sharding and all the outputs with disable_manual_sharding.

Added a new test file because the original test file is not SPMD aware.

Test Plan:
PJRT_DEVICE=TPU python test/test_pallas_spmd.py

### CODE REVIEW

**Review by JackCaoG — APPROVED** [review]
lgtm, minor comments

File: test/test_pallas_spmd.py [hunk 1]
```diff
@@ -0,0 +1,114 @@
...
+import torch_xla.distributed.spmd as xs
+from torch_xla import runtime as xr
+from torch_xla._internal import tpu
+
+if xr.device_type() == 'TPU':
+  from torch_xla.experimental.custom_kernel import jax_import_guard
+  jax_import_guard()
+  import jax
+  import jax.numpy as jnp
+  from jax.experimental import pallas as pl
```
[the remark this answers is not in the document]
Comment (alanwaketan) [reply test/test_pallas_spmd.py:19]: Is python import global?

File: test/test_pallas_spmd.py [hunk 2]
```diff
@@ -0,0 +1,114 @@
...
+    attn_output = attn_weight @ v
+    return attn_output
+
+  @unittest.skipIf(xr.device_type() != 'TPU' or tpu.version() < 3,
+                   "This test only works on TPUv3+.")
+  def test_flash_attention_spmd_data_parallel(self):
+    jax.config.update('jax_default_matmul_precision', jax.lax.Precision.HIGHEST)
+    from torch_xla.experimental.custom_kernel import flash_attention
+
+    xr.use_spmd()
```
[the remark this answers is not in the document]
Comment (alanwaketan) [reply test/test_pallas_spmd.py:36]: I probably can call this in `main` as well. The setup class seems overkilled for this.

File: test/test_pallas_spmd.py [hunk 3]
```diff
@@ -0,0 +1,114 @@
...
+  def _attention(self, q, k, v):
+    attn_weight = q @ k.transpose(-2, -1)
+    attn_weight = nn.functional.softmax(attn_weight, dim=-1)
+    attn_output = attn_weight @ v
+    return attn_output
+
+  @unittest.skipIf(xr.device_type() != 'TPU' or tpu.version() < 3,
+                   "This test only works on TPUv3+.")
+  def test_flash_attention_spmd_data_parallel(self):
+    jax.config.update('jax_default_matmul_precision', jax.lax.Precision.HIGHEST)
```
Comment (jonb377) [inline test/test_pallas_spmd.py:33]: Does this impact the resulting kernel?

File: torch_xla/experimental/custom_kernel.py [hunk 4]
```diff
@@ -184,15 +185,29 @@ class FlashAttention(torch.autograd.Function):
   }
 
   @staticmethod
-  def forward(ctx, q, k, v, causal=False):
+  def forward(ctx, q, k, v, causal=False, sharding_spec=None, mesh=None):
```
Comment (jonb377) [inline torch_xla/experimental/custom_kernel.py:188]: nit: `sharding_spec` -> `partition_spec`?

**Review by jonb377 — APPROVED** [review]
Awesome stuff Jiewen!

File: test/test_pallas_spmd.py (hunk 3 above)
Comment (alanwaketan) [reply test/test_pallas_spmd.py:33]: Yea.

**Comment by alanwaketan** [comment]
Thanks Jon and Jack for the reviews.

**Pull request merged**